Feature/fpca_regression

Ddelval opened the pull request:
Add the FPCA regression estimator

vnmabus reviewed (changes requested):
I think this functionality is implemented in the package `fda.usc`. We should have a test verifying that we obtain the same results.

Ddelval commented:
> I think this functionality is implemented in the package `fda.usc`. We should have a test verifying that we obtain the same results.

I have (finally) added this test. However, while going back and forth between python and R, I stumbled upon a python packet that provides a simple way to execute R code from within python. I am sceptical regarding whether or not this could be useful for the tests. On the one hand, it would ensure that the tests are testing against the correct values while forcing us to include the R source used to obtain the test reference data. On the other hand, it would make the tests considerably slower and require modifications to the GitHub actions.

I think the timing tradeoff is too significant (I measured a slowdown of around 6x). Nevertheless, I thought I might just mention it just in case it comes in handy at some point. The following snippet generates the testing data for the test I just added.
```
from rpy2 import robjects

Rresult = robjects.r('''
library("fda.usc")
data(tecator)
# Fit the regression model with the first 129 observations
x=tecator$absorp.fdata[1:129,]
y=tecator$y$Fat[1:129]
res2=fregre.pc(x,y,l=1:10)

# Predict the response for the remaining observations
n = length(tecator$y$Fat)
xnew=tecator$absorp.fdata[130:n,]
result = predict(res2, xnew)
names(result) = NULL

# Output the predicted values
paste(
    round(result,8),
    collapse = ", "
)
''')
r_predictions = np.array(Rresult[0].split(", "), dtype=np.float64)
```

vnmabus commented on skfda/ml/regression/_fpca_regression.py:
```diff
@@ -0,0 +1,150 @@
+
+        self.n_components_ = self.n_components
+        self.components_ = self._fpca.components_
+        self.coefficients_ = self._linear_model.coef_
```
Why don't you use the same name?

vnmabus commented on skfda/ml/regression/_fpca_regression.py:
```diff
@@ -0,0 +1,150 @@
+    Attributes:
+        n\_components\_: Number of principal components used.
+        components\_: Principal components.
+        cofficents\_: Coefficients of the linear regression model.
```
There is a typo here.
```suggestion
        coefficients\_: Coefficients of the linear regression model.
```

vnmabus commented on skfda/ml/regression/_fpca_regression.py:
```diff
@@ -0,0 +1,150 @@
+    def __init__(
+        self,
+        n_components: int = 2,
+        intercept: bool = True,
```
Why not `fit_intercept`?

vnmabus commented on skfda/ml/regression/_fpca_regression.py:
```diff
@@ -0,0 +1,150 @@
+    def fit(
+        self,
+        X: FData,
+        y: np.ndarray,
```
Use a more specific type if possible.

vnmabus commented on skfda/ml/regression/_fpca_regression.py:
```diff
@@ -0,0 +1,150 @@
+    def predict(
+        self,
+        X: FData,
+    ) -> np.ndarray:
```
Use a more specific type.

vnmabus commented on skfda/representation/basis/_custom_basis.py:
```diff
@@ -195,6 +195,22 @@ def _evaluate(
+            coefficients = self.fdata.coefficients
+            return coefficients @ basis_gram @ coefficients.T
+
+        from ...misc import inner_product_matrix
```
Do the import at the beginning of the function (just after the docstring).

vnmabus commented on skfda/tests/test_fpca_regression.py:
```diff
@@ -0,0 +1,212 @@
+        """Check that the results obtained are similar to those of fda.usc.
+
+        Results obtained from fda.usc with the following R code
+            library("fda.usc")
```
Were you able to compare results with refund too?

Ddelval replied:
This is the issue that I opened: https://github.com/refunders/refund/issues/102.
It doesn't seem like we'll be getting any solutions from their side anytime soon.

vnmabus reviewed (changes requested):
Remember to also add the class to the API reference documentation.

vnmabus commented on skfda/representation/basis/_custom_basis.py:
```diff
@@ -195,6 +195,23 @@ def _evaluate(
+        """
+        from ...misc import inner_product_matrix
+
+        if isinstance(self.fdata, FDataBasis):
```
Why is this specialized for `FDataBasis`?

Ddelval replied:
Leftover from the previous implementation. Fixed now.

vnmabus commented on skfda/ml/regression/_fpca_regression.py:
```diff
@@ -0,0 +1,152 @@
+            Target values.
+
+        """
+        check_is_fitted(self, ["_fpca", "_linear_model"])
```
I think you do not need to pass the second argument to this function.
